Port all setup.py files to setuptools

tiran opened the pull request:
PREVIEW, please don't merge yet.

This is my take on FreeIPA's setup.py files. I have moved all common code and constants into ipasetup.py.in and converted all setup.py.in to setup.py. Setup now uses setuptools instead of distutils. I had to move ipa.1 man page to client/man because setuptools does no longer support data_files.

Signed-off-by: Christian Heimes <cheimes@redhat.com>

mbasti-rh commented on ipasetup.py.in:
```diff
@@ -0,0 +1,64 @@
+    download_url="http://www.freeipa.org/page/Downloads",
+    platforms=["Linux", "Solaris", "Unix"],
+    classifiers=[
+        "Development Status :: 4 - Beta",
```
Maybe we can change this to anything else than beta :D

mbasti-rh commented:
This WIP works for me, I like that we get rid of setup.py.in files.

I'm looking forward to final version

Please fix PEP8 reported error and my inline comments

tiran commented:
@mbasti-rh I have removed more hacks and made each setup.py even simpler.

jcholast commented on .gitignore:
```diff
@@ -41,6 +41,9 @@ freeipa2-dev-doc
 /RELEASE
 /rpmbuild/
+# Build
+ipasetup.py
```
Why is `ipasetup.py` ignored?

tiran replied:
Makefile autogenerates the file from ipasetup.py.in.

jcholast replied:
*slaps forehead* Right.

jcholast commented on Makefile:
```diff
@@ -161,23 +161,25 @@ test:
+	for directory in ipaclient ipalib ipaplatform ipapython ipatests; do \
+	    pushd $${directory} ; \
+	    cp ../ipasetup.py . ; \
+	    $(PYTHON) setup.py egg_info ; \
```
IMO it would be better to run `setup.py` with `PYTHONPATH` set properly, rather than copy `ipasetup.py` all the time.

jcholast replied:
Alternatively, if `sys.path` was updated appropriately in each `setup.py`, we would get rid of copying _and_ allow chdir & run `setup.py`.

jcholast commented on client/man/Makefile.am:
```diff
@@ -10,7 +10,8 @@ man1_MANS = 				\
 		ipa-certupdate.1	\
-		ipa-join.1
+		ipa-join.1		\
+		ipa.1
```
Either put the `ipa.1` packaging change in a separate commit, or don't do it at all (there must be some way to package custom file with `setuptool`, right?)

tiran replied:
No, setuptools no longer supports non-package data. It can no longer install files to ```PREFIX/share```, ```PREFIX/man``` or any other directory except for scripts (```PREFIX/share```) and package data.

mbasti-rh commented:
I cannot reinstall packages using your commits

```
dnf reinstall <IPA*>.rpm
...
Error: Transaction check error:
  file /usr/lib/python2.7/site-packages/ipalib-4.4.90-py2.7.egg-info from install of python2-ipalib-4.4.90-0.fc24.noarch conflicts with file from package python2-ipalib-4.4.90-0.fc24.noarch
  file /usr/lib/python2.7/site-packages/ipaplatform-4.4.90-py2.7.egg-info from install of python2-ipalib-4.4.90-0.fc24.noarch conflicts with file from package python2-ipalib-4.4.90-0.fc24.noarch
  file /usr/lib/python2.7/site-packages/ipapython-4.4.90-py2.7.egg-info from install of python2-ipalib-4.4.90-0.fc24.noarch conflicts with file from package python2-ipalib-4.4.90-0.fc24.noarch
  file /usr/lib/python2.7/site-packages/ipaclient-4.4.90-py2.7.egg-info from install of python2-ipaclient-4.4.90-0.fc24.noarch conflicts with file from package python2-ipaclient-4.4.90-0.fc24.noarch
  file /usr/lib/python3.5/site-packages/ipalib-4.4.90-py3.5.egg-info from install of python3-ipalib-4.4.90-0.fc24.noarch conflicts with file from package python3-ipalib-4.4.90-0.fc24.noarch
  file /usr/lib/python3.5/site-packages/ipaplatform-4.4.90-py3.5.egg-info from install of python3-ipalib-4.4.90-0.fc24.noarch conflicts with file from package python3-ipalib-4.4.90-0.fc24.noarch
  file /usr/lib/python3.5/site-packages/ipapython-4.4.90-py3.5.egg-info from install of python3-ipalib-4.4.90-0.fc24.noarch conflicts with file from package python3-ipalib-4.4.90-0.fc24.noarch
  file /usr/lib/python3.5/site-packages/ipaclient-4.4.90-py3.5.egg-info from install of python3-ipaclient-4.4.90-0.fc24.noarch conflicts with file from package python3-ipaclient-4.4.90-0.fc24.noarch
  file /usr/lib/python3.5/site-packages/ipatests-4.4.90-py3.5.egg-info from install of python3-ipatests-4.4.90-0.fc24.noarch conflicts with file from package python3-ipatests-4.4.90-0.fc24.noarch
  file /usr/lib/python2.7/site-packages/ipatests-4.4.90-py2.7.egg-info from install of python2-ipatests-4.4.90-0.fc24.noarch conflicts with file from package python2-ipatests-4.4.90-0.fc24.noarch

```

tiran commented:
@mbasti-rh I can't reproduce your problem. dnf reinstall works for me. I tested both upgrades from 4.4.2 and reinstall of my RPMs.

jcholast commented on MANIFEST.in:
```diff
@@ -1,2 +1 @@
-include COPYING TODO lite-server.py
-include tests/*/*.py
+include COPYING lite-server.py
```
Where is this manifest file used? Can't we just drop it? I think it at least should not include lite-server.py.

tiran commented:
The PR is no longer provisional and ready-to-merge.

jcholast commented on setup.py:
```diff
@@ -23,15 +23,15 @@
     version=ipalib.__version__,
-    license='GPLv3+',
-    url='http://freeipa.org/',
+    scripts=['ipa'],
```
The `ipa` script is in fact a client script, so I guess it shouldn't be packaged with `ipaserver`.

jcholast commented:
LGTM, but shouldn't we also move `/setup.py` to `/ipaserver/setup.py`?

tiran commented:
I have moved /setup.py into /ipaserver/, fixed some make distclean targets (they were still removing setup.py) and made /ipaclient/setup.u install the ipa script. make rpms is passing.

pspacek commented:
Jenkins does not complain. ACK.

pvoborni commented:
Fixed upstream
master:
https://fedorahosted.org/freeipa/changeset/4cd83fb51cc35a2ba7773b62a7aa8d295a1e1e4a
https://fedorahosted.org/freeipa/changeset/e12a70a8b1aa5daac4b8ab7ac681b09f664a8357
